### PR TITLE
[RFC] Menu source

### PR DESCRIPTION
**Pull request by tjdevries**
Hey Shougo,

I was working on a menu source just so that I could use it. It's still a work in progress at this point, but I figured I'd just make the PR.

I do have file candidates working currently. They use the same format as the old `unite_source_menu_menus` dictionary.

No rush on any of this. I know you have a lot of TODO's for the project.

Thanks!

### CODE REVIEW

File: rplugin/python3/denite/source/menu.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
+# ============================================================================
+# FILE: line.py
```
Comment (Shougo) [inline rplugin/python3/denite/source/menu.py]: menu.py

File: rplugin/python3/denite/source/menu.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+        self.name = 'menu'
+        self.kind = 'jump_list'
+
+        # self.matchers = []
+        # self.sorters = []
+
+    def on_init(self, context):
+        # TODO: Could also check for unite_source_menu_menus?
+        context['__menus'] = self.vim.vars['denite_source_menu_menus']
```
Comment (Shougo) [inline rplugin/python3/denite/source/menu.py]: It should be `context['vars'].get('menus', default)`.

File: rplugin/python3/denite/source/menu.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+    def gather_candidates(self, context):
+        lines = []
+
+        menu_context = [option for option in context['sources'] if option['name'] == self.name][0]
+
+        if menu_context == []:
+            # Exit quickly
+            return []
+        else:
```
Comment (Shougo) [inline rplugin/python3/denite/source/menu.py]: It should be:

```python
      if menu_context == []:
            # Exit quickly
            return []

     if menu_context['args']:
...
```

File: rplugin/python3/denite/source/menu.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        # self.sorters = []
+
+    def on_init(self, context):
+        # TODO: Could also check for unite_source_menu_menus?
+        context['__menus'] = self.vim.vars['denite_source_menu_menus']
+
+    def gather_candidates(self, context):
+        lines = []
+
+        menu_context = [option for option in context['sources'] if option['name'] == self.name][0]
```
Comment (Shougo) [inline rplugin/python3/denite/source/menu.py]: It should be:

```python
menu_context = context['args']
```

**Comment by Shougo** [comment]
Please fix the lint errors.

File: rplugin/python3/denite/source/menu.py (hunk 4 above)
Comment (tjdevries) [inline rplugin/python3/denite/source/menu.py]: This change didn't work for me. I left it the way it is for now. I am getting the file candidates to pop up with the original way I had it, but not with the other way.

**Comment by tjdevries** [comment]
To test this, I'm just running:

```
nvim -u test_rc.vim -c "Denite menu:init_vim"
```

with test_rc.vim being:

```vim
scriptencoding utf-8

call plug#begin('~/test/denite/.plugged')
Plug '~/Git/denite.nvim'
call plug#end()

let g:my_var = 'DENITE IS COOL'

let g:denite_source_menu_menus = {}

let g:denite_source_menu_menus.init_vim = {
        \ 'description': 'Edit your important init.vim information'
        \ }
let g:denite_source_menu_menus.init_vim.file_candidates = [
        \ ['init', '~/.config/nvim/init.vim'],
        \ ['init/', '~/.config/nvim/init/'],
        \ ['basic configuration', '~/.config/nvim/init/01-basic_configuration.vim'],
        \ ['plugins', '~/.config/nvim/init/02-plugins.vim'],
        \ ['↳airline', '~/.config/nvim/init/05-airline.vim'],
        \ ['↳deoplete', '~/.config/nvim/init/05-deoplete.vim'],
        \ ['↳neomake', '~/.config/nvim/init/05-neomake.vim'],
        \ ['↳startify', '~/.config/nvim/init/05-startify.vim'],
        \ ['↳denite', '~/.config/nvim/init/05-denite.vim'],
        \ ]

let g:denite_source_menu_menus.zsh = {
        \ 'description': 'Edit your import zsh configuration'
        \ }
let g:denite_source_menu_menus.zsh.file_candidates = [
        \ ['zshrc', '~/.config/zsh/.zshrc'],
        \ ['include', '~/.config/zsh/include/'],
        \ ['↳aliases', '~/.config/zsh/include/aliases.zsh'],
        \ ['↳autojump', '~/.config/zsh/include/autojump.zsh'],
        \ ['↳functions', '~/.config/zsh/include/functions.zsh'],
        \ ['↳git', '~/.config/zsh/include/git.zsh'],
        \ ]
```

File: rplugin/python3/denite/source/menu.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+        self.name = 'menu'
+        self.kind = 'jump_list'
+
+        # self.matchers = []
+        # self.sorters = []
+
+    def on_init(self, context):
+        # TODO: Could also check for unite_source_menu_menus?
+        context['__menus'] = context.get('menus',
```
Comment (Shougo) [inline rplugin/python3/denite/source/menu.py]: I think we don't  add the global variable like `g:denite_source_menu_menus`.

File: rplugin/python3/denite/source/menu.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+                                         self.vim.vars.get(
+                                             'denite_source_menu_menus',
+                                             {}
+                                         ))
+
+    def gather_candidates(self, context):
+        lines = []
+
+        menu_context = [option for option in context['sources']
```
Comment (Shougo) [inline rplugin/python3/denite/source/menu.py]: Please use `context['args']` instead.
Ref: file_rec.py

**Comment by Shougo** [comment]
I have added `denite#custom#var()`.  Please use it instead of `g:denite_source_menu_menus`.

```vim
let g:denite_source_menu_menus = {}

let g:denite_source_menu_menus.init_vim = {
        \ 'description': 'Edit your important init.vim information'
        \ }
let g:denite_source_menu_menus.init_vim.file_candidates = [
        \ ['init', '~/.config/nvim/init.vim'],
        \ ['init/', '~/.config/nvim/init/'],
        \ ['basic configuration', '~/.config/nvim/init/01-basic_configuration.vim'],
        \ ['plugins', '~/.config/nvim/init/02-plugins.vim'],
        \ ['↳airline', '~/.config/nvim/init/05-airline.vim'],
        \ ['↳deoplete', '~/.config/nvim/init/05-deoplete.vim'],
        \ ['↳neomake', '~/.config/nvim/init/05-neomake.vim'],
        \ ['↳startify', '~/.config/nvim/init/05-startify.vim'],
        \ ['↳denite', '~/.config/nvim/init/05-denite.vim'],
        \ ]

let g:denite_source_menu_menus.zsh = {
        \ 'description': 'Edit your import zsh configuration'
        \ }
let g:denite_source_menu_menus.zsh.file_candidates = [
        \ ['zshrc', '~/.config/zsh/.zshrc'],
        \ ['include', '~/.config/zsh/include/'],
        \ ['↳aliases', '~/.config/zsh/include/aliases.zsh'],
        \ ['↳autojump', '~/.config/zsh/include/autojump.zsh'],
        \ ['↳functions', '~/.config/zsh/include/functions.zsh'],
        \ ['↳git', '~/.config/zsh/include/git.zsh'],
        \ ]
```

File: rplugin/python3/denite/source/menu.py (hunk 5 above)
Comment (Shougo) [inline rplugin/python3/denite/source/menu.py]: To add `g:unite_source_menus` support is not bad.
But it supports Function type.
denite cannot support it.

**Comment by tjdevries** [comment]
Okay, I have changed it so that now it uses `denite#custom#var()`. I like this approach better.

I still have several other TODOs.

File: rplugin/python3/denite/source/menu.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+    def __init__(self, vim):
+        Base.__init__(self, vim)
+
+        self.name = 'menu'
+        self.kind = 'jump_list'
+
+        # self.matchers = []
+        # self.sorters = []
+
+        self.dict_key = '__menus'
```
Comment (Shougo) [inline rplugin/python3/denite/source/menu.py]: It should be:
```python
self.vars = {
  'menus': {}
}
```

File: rplugin/python3/denite/source/menu.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+
+    def on_init(self, context):
+        context[self.dict_key] = context['custom']['source']['menu']['vars']\
```
Comment (Shougo) [inline rplugin/python3/denite/source/menu.py]: You don't have to set the context.
self.vars is overwritten by `denite#custom#var()`.

Comment (tjdevries) [inline rplugin/python3/denite/source/menu.py]: :+1: This is an awesome feature :D

**Comment by tjdevries** [comment]
I like the `self.vars`. Much cleaner. I can't implement command candidates or other types of candidates yet, but the rest seems to be working for me.

Thanks for all the feedback!

**Comment by Shougo** [comment]
Thanks.

> I like the self.vars. Much cleaner. I can't implement command candidates or other types of candidates yet, but the rest seems to be working for me.

I will implement command kind.
Please wait.

**Comment by tjdevries** [comment]
:+1: Sounds good to me.

**Comment by tjdevries** [comment]
How would I do a mix of `jump_list` (for file candidates) and `command` (for command candidates) within the same source?

**Comment by Shougo** [comment]
> How would I do a mix of jump_list (for file candidates) and command (for command candidates) within the same source?

Please use 'kind' key in the candidates.
`source.kind` is just default kind.

**Comment by tjdevries** [comment]
Excellent. Thanks

**Comment by Shougo** [comment]
This feature is not implemented in ctrlp or other fuzzy finders :-)

**Comment by tjdevries** [comment]
Let's say I wanted to have the `command` do something simple, like `vnew`, what would my `action__command` have to be? I seem to not be able to get it to work correctly. Sorry if I'm missing something obvious.

**Comment by Shougo** [comment]
What's the your problem?
Please describe it.

**Comment by tjdevries** [comment]
If I have a command candidate that where the `action__command` is set to be: `vnew`, then I get this error in neovim:

```[denite] Traceback (most recent call last):                                                                          
[denite]   File "/home/tj/Git/denite.nvim/rplugin/python3/denite/ui/default.py", line 51, in start                   
[denite]     self.input_loop(context)                                                                                
[denite]   File "/home/tj/Git/denite.nvim/rplugin/python3/denite/ui/default.py", line 146, in input_loop             
[denite]     ret = func(context)                                                                                     
[denite]   File "/home/tj/Git/denite.nvim/rplugin/python3/denite/ui/default.py", line 171, in do_action              
[denite]     self.__denite.do_action(context, kind, 'default', [candidate])                                          
[denite]   File "/home/tj/Git/denite.nvim/rplugin/python3/denite/denite.py", line 171, in do_action                  
[denite]     func(context)                                                                                           
[denite]   File "/home/tj/Git/denite.nvim/rplugin/python3/denite/kind/command.py", line 19, in action_default        
[denite]     self.vim.call(target['action__command'])                                                                
[denite]   File "/home/tj/.local/lib/python3.5/site-packages/neovim/api/nvim.py", line 213, in call                  
[denite]     return self.request('vim_call_function', name, args, **kwargs)                                          
[denite]   File "/home/tj/.local/lib/python3.5/site-packages/neovim/api/nvim.py", line 129, in request               
[denite]     res = self._session.request(name, *args, **kwargs)                                                      
[denite]   File "/home/tj/.local/lib/python3.5/site-packages/neovim/msgpack_rpc/session.py", line 98, in request     
[denite]     raise self.error_wrapper(err)                                                                           
[denite] neovim.api.nvim.NvimError: b'Vim:E117: Unknown function: vnew'  ```

I'm not sure how to avoid that error.

**Comment by Shougo** [comment]
It is denite bug.  Fixed.

**Comment by tjdevries** [comment]
Now I am getting this:

```
[denite] Traceback (most recent call last):                                                                          
[denite]   File "/home/tj/Git/denite.nvim/rplugin/python3/denite/ui/default.py", line 51, in start                   
[denite]     self.input_loop(context)                                                                                
[denite]   File "/home/tj/Git/denite.nvim/rplugin/python3/denite/ui/default.py", line 146, in input_loop             
[denite]     ret = func(context)                                                                                     
[denite]   File "/home/tj/Git/denite.nvim/rplugin/python3/denite/ui/default.py", line 171, in do_action              
[denite]     self.__denite.do_action(context, kind, 'default', [candidate])                                          
[denite]   File "/home/tj/Git/denite.nvim/rplugin/python3/denite/denite.py", line 171, in do_action                  
[denite]     func(context)                                                                                           
[denite]   File "/home/tj/Git/denite.nvim/rplugin/python3/denite/kind/command.py", line 19, in action_default        
[denite]     self.vim.execute(target['action__command'])                                                             
[denite] AttributeError: 'Nvim' object has no attribute 'execute'                                                    
[denite] An error has occurred. Please execute :messages command.
```

**Comment by Shougo** [comment]
Woops.  Fixed.

**Comment by Shougo** [comment]
Please add the documentation for the source.

**Comment by carlitux** [comment]
@tjdevries @Shougo there is a way to get all function from neovim and plugins? so we don't need to define the menus or maybe create another source for that.

**Comment by tjdevries** [comment]
@carlitux, I'm not sure what you're saying exactly. These are custom menus that a user creates to help them with tasks that are unique to them. For example, if you look above at the menu creation, those are my personal menus I like to have to quickly move between various configuration files I have. I'm not sure what getting all the function from neovim and plugins would do.

**Comment by carlitux** [comment]
@tjdevries if you have used sublime, a user can access all commands from fuzzy finder... I think that could be a good feature.

**Comment by tjdevries** [comment]
I think that would be a different type of source.

It is a good idea though! Perhaps I will look into that after this source is finished. I do not think they would overlap in functionality though. These menus are for small subsets of specific tasks.

For example, I have mapped:

```vim
nnoremap <leader>en :Denite menu:init_vim
```

in my `init.vim` so that whenever I want to quickly edit something in my configuration, I just use that mapping and I'm there. I'm not sure your fuzzy finder idea would provide the granularity I'd like.

**Comment by carlitux** [comment]
@tjdevries yes. I think same, too. Now I understood what your source does. Thanks! I will try to create that source if I have free time

**Comment by tjdevries** [comment]
@carlitux excellent :smile: Make sure to post it as a PR. I'd be happy to help if I'm able.

**Comment by Shougo** [comment]
> @tjdevries @Shougo there is a way to get all function from neovim and plugins? so we don't need to define the menus or maybe create another source for that.

You need new sources(command and function) for it.
unite implements both sources.  You should check it.

**Comment by tjdevries** [comment]
@Shougo, Let me know what else you want fixed before this is ready. Works for my machine and does what the old unite menu source did (at least for me).

Thanks.

File: doc/denite.txt [hunk 9]
```diff
@@ -256,6 +278,18 @@ grep		Gather grep results and nominates them.
...
 
+						*denite-source-menu*
+menu		Gather custom menus.
+
+		Source arguments:
+		1. The desired menu
+
+		Source custom variables:
+		menus		dictionary containing menus in key value
+				pairs (See |denite-examples|).
```
Comment (Shougo) [inline doc/denite.txt]: Please add `unite_source_menu_compatibility` description.

Comment (Shougo) [inline doc/denite.txt]: It supports only List type.
It should be documented.

File: rplugin/python3/denite/source/menu.py [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+
+        lines = []
+        menus = self.vars['menus']
+        args = context['args']
+
+        if args:
+            # Loop through each menu option
+            for menu in args:
+                # If a menu doesn't exist, just continue gracefully
+                if menu not in menus.keys():
```
Comment (Shougo) [inline rplugin/python3/denite/source/menu.py]: It should be

```python
if menu in menus:
```

**Comment by Shougo** [comment]
> @Shougo, Let me know what else you want fixed before this is ready. Works for my machine and does what the old unite menu source did (at least for me).

Thanks.  I have reviewed.

**Comment by tjdevries** [comment]
@Shougo I think I have made the changes. I'm not sure why it's not showing that I changed those places, perhaps it is because I rebased against master.

Anyways, let me know if you want anything else updated. I think You have access to just change things from github if you'd like with the new pull request features.

**Comment by Shougo** [comment]
Merged.

**Pull request merged**